drivers/led_strip/ws2812_rpi_pico_pio.c

```diff
@@ -0,0 +1,248 @@
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_rpi_pico_pio, CONFIG_LED_STRIP_LOG_LEVEL);
+
+#define DT_DRV_COMPAT worldsemi_ws2812_rpi_pico_pio
+
+struct ws2812_led_strip_data {
+	uint32_t sm;
+};
+
+struct ws2812_led_strip_config {
+	const struct device *piodev;
+	uint32_t output_pin;
+	uint8_t num_colors;
+	uint32_t frequency;
+	const uint8_t *const color_mapping;
+	uint16_t reset_delay;
+	uint32_t cycles_per_bit;
+};
+
+struct ws2812_rpi_pico_pio_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *const pcfg;
+	struct pio_program program;
+};
+
+static int ws2812_led_strip_sm_init(const struct device *dev)
+{
+	const struct ws2812_led_strip_config *config = dev->config;
+	const float clkdiv =
+		sys_clock_hw_cycles_per_sec() / (config->cycles_per_bit * config->frequency);
+	pio_sm_config sm_config = pio_get_default_sm_config();
+	PIO pio;
+	int sm;
+
+	if (!device_is_ready(config->piodev)) {
+		LOG_ERR("%s: PIO device not ready", dev->name);
+		return -ENODEV;
+	}
+
+	pio = pio_rpi_pico_get_pio(config->piodev);
+
+	sm = pio_claim_unused_sm(pio, false);
+	if (sm < 0) {
+		return -EINVAL;
+	}
+
+	sm_config_set_sideset(&sm_config, 1, false, false);
+	sm_config_set_sideset_pins(&sm_config, config->output_pin);
+	sm_config_set_out_shift(&sm_config, false, true, (config->num_colors == 4 ? 32 : 24));
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, clkdiv);
+	pio_sm_set_consecutive_pindirs(pio, sm, config->output_pin, 1, true);
+	pio_sm_init(pio, sm, -1, &sm_config);
+	pio_sm_set_enabled(pio, sm, true);
+
+	return sm;
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_led_strip_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
+
+static int ws2812_led_strip_update_rgb(const struct device *dev, struct led_rgb *pixels,
+				       size_t num_pixels)
+{
+	const struct ws2812_led_strip_config *config = dev->config;
+	struct ws2812_led_strip_data *data = dev->data;
+	PIO pio = pio_rpi_pico_get_pio(config->piodev);
+
+	for (size_t i = 0; i < num_pixels; i++) {
+		uint32_t color = 0;
+
+		for (size_t j = 0; j < config->num_colors; j++) {
+			switch (config->color_mapping[j]) {
+			/* White channel is not supported by LED strip API. */
+			case LED_COLOR_ID_WHITE:
+				color |= 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color |= pixels[i].r << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_GREEN:
+				color |= pixels[i].g << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_BLUE:
+				color |= pixels[i].b << (8 * (2 - j));
+				break;
+			}
+		}
+
+		pio_sm_put_blocking(pio, data->sm, color << (config->num_colors == 4 ? 0 : 8));
+	}
+
+	ws2812_led_strip_reset_delay(config->reset_delay);
+
+	return 0;
+}
+
+static int ws2812_led_strip_update_channels(const struct device *dev, uint8_t *channels,
+					    size_t num_channels)
+{
+	LOG_ERR("update_channels not implemented");
```

Comment:
I would decrease the log level to DBG or even remove it completely.

Reply:
Changed it to DBG.